fenix/app/src/main/java/org/mozilla/fenix/browser/tabstrip/TabsStrip.kt

```diff
@@ -0,0 +1,371 @@
+import androidx.compose.foundation.lazy.items
+import androidx.compose.foundation.lazy.rememberLazyListState
+import androidx.compose.foundation.shape.CircleShape
+import androidx.compose.foundation.systemGestureExclusion
+import androidx.compose.material.Icon
+import androidx.compose.material.IconButton
+import androidx.compose.material.Text
+import androidx.compose.runtime.Composable
+import androidx.compose.runtime.LaunchedEffect
+import androidx.compose.runtime.getValue
+import androidx.compose.ui.Alignment
+import androidx.compose.ui.Modifier
+import androidx.compose.ui.draw.clip
+import androidx.compose.ui.graphics.asImageBitmap
+import androidx.compose.ui.res.painterResource
+import androidx.compose.ui.res.stringResource
+import androidx.compose.ui.text.style.TextOverflow
+import androidx.compose.ui.tooling.preview.Devices
+import androidx.compose.ui.tooling.preview.Preview
+import androidx.compose.ui.tooling.preview.PreviewParameter
+import androidx.compose.ui.tooling.preview.PreviewParameterProvider
+import androidx.compose.ui.unit.coerceIn
+import androidx.compose.ui.unit.dp
+import mozilla.components.lib.state.ext.observeAsState
+import org.mozilla.fenix.R
+import org.mozilla.fenix.components.components
+import org.mozilla.fenix.theme.FirefoxTheme
+import org.mozilla.fenix.theme.Theme
+
+private val minTabStripItemWidth = 160.dp
+private val maxTabStripItemWidth = 280.dp
+private val tabStripIconSize = 24.dp
+private val spaceBetweenTabs = 4.dp
+private val tabStripStartPadding = 8.dp
+
+/**
+ * Top level composable for the tabs strip.
+ *
+ * @param onHome Whether or not the tabs strip is in the home screen.
+ * @param onAddTabClick Invoked when the add tab button is clicked.
+ * @param onLastTabClose Invoked when the last tab is closed.
+ * @param onSelectedTabClick Invoked when a tab is selected.
+ */
+@Composable
+fun TabsStrip(
+    onHome: Boolean = false,
+    onAddTabClick: () -> Unit,
+    onLastTabClose: () -> Unit = {},
+    onSelectedTabClick: () -> Unit = {},
+) {
+    val browserStore = components.core.store
+    val tabsUseCases = components.useCases.tabsUseCases
+    val appStore = components.appStore
+    val isPrivateMode by appStore.observeAsState(false) { it.mode.isPrivate }
+    val state by browserStore.observeAsState(TabsStripState.initial) {
+        it.toTabsStripState(onHome = onHome, isPrivateMode = isPrivateMode)
+    }
+
+    TabStripContent(
+        state = state,
+        onAddTabClick = onAddTabClick,
+        onCloseTabClick = {
+            if (state.tabs.size == 1) {
+                onLastTabClose()
+            }
+            tabsUseCases.removeTab(it)
+        },
+        onSelectedTabClick = {
+            tabsUseCases.selectTab(it)
+            onSelectedTabClick()
+        },
+    )
+}
+
+@Composable
+private fun TabStripContent(
+    state: TabsStripState,
+    onAddTabClick: () -> Unit,
+    onCloseTabClick: (id: String) -> Unit,
+    onSelectedTabClick: (id: String) -> Unit,
+) {
```

Comment:
one downside to the composable + content pattern is that the previews will no longer test the store observations. Have you considered just passing in `BrowserStore` and `AppStore` as explicit parameters instead?

The `tabsUseCases` could be hoisted as well and be incorporated into click lambdas. From a cursory look, the `tabsUseCases` might be able to be constructed within previews, since their only dependency is `BrowserStore`, so we could have fully functional previews using our stores + use cases without needing to use `components` at this level!

Reply:
Yeah I did consider this and initially had it like that but that was leading to duplication of calls to tabsUsecases functions in both the places TabStrip is used. This `TabStrip` top level composable acts like a controller/fragment and the `Content` composable is the UI that simply gets the data. So the parent fragments don't have to setup similarly everywhere. We can extract another top level composable that uses components and this composable (can be made private) can simply observe but I'm not sure if there's benefit to testing observation in previews since the mapper is tested and observation code is tested, and the content composable gets the simplified data model.

Reply:
I just investigated a potential option for this that combines both approaches:

```
@Composable
fun TabsStrip(
    appStore: AppStore = components.appStore,
    browserStore: BrowserStore = components.core.store,
    tabsUseCases: TabsUseCases = components.useCases.tabsUseCases,
    onHome: Boolean = false,
    onAddTabClick: () -> Unit,
    onLastTabClose: () -> Unit = {},
    onSelectedTabClick: () -> Unit = {},
)
```
```
@Composable
@Preview
private fun TabStripPreview() {
    val browserStore = BrowserStore()

    FirefoxTheme {
        Box(
            modifier = Modifier
                .fillMaxWidth()
                .height(56.dp),
            contentAlignment = Alignment.Center,
        ) {
            TabsStrip(
                appStore = AppStore(),
                browserStore = browserStore,
                tabsUseCases = TabsUseCases(browserStore),
                onAddTabClick = {
                    val tab = createTab(
                        url = "www.example.com",
                    )
                    browserStore.dispatch(TabListAction.AddTabAction(tab))
                },
                onLastTabClose = {},
                onSelectedTabClick = {},
            )
        }
    }
}
```

Giving us:

https://github.com/mozilla-mobile/firefox-android/assets/87384386/8b6012c9-8c60-4568-9c66-7eeff99ee691

Reply:
And we could still keep the existing `content` composable and the parameterized previews if we wanted.

Reply:
Thanks for writing down your thoughts and the video! And the setup cost isn't much really too. That wasn't the best reasoning from my side. After thinking more, I agree this is much better. I've updated and removed the default params for functions, they should be explicit too.